### PR TITLE
Upgrade node to 16.13

### PR DESCRIPTION
**Pull request by coltoneshaw**
#### Summary
Started trying to build a plugin and hit a few roadblocks with npm versions and the packages. Below is updating everything to work from the start on the latest node/npm version to hopefully make starting a plugin easier!
 
- Increased the versions to the latest across all the possible packages. React version remains at v16.14.0 because `enzyme` does not have React v17 support officially. 
- Rebuilt the `package-lock.json` file with the new versions.
- Updated the docs to include a version number for `node` / `npm`.
- Made a few small changes to the webpack config to clean up the `isDev` conditional.
- Added the `babel` plugin to eslint to support the latest webpack.
- removed the Mattermost-redux dependency
- Built with Node 16.13 to maintain LTS

### CODE REVIEW

**Comment by coltoneshaw** [comment]
Okay, all tests pass locally. The exception is `npm run build` which seems to require `export NODE_OPTIONS=--openssl-legacy-provider` before it to properly run.  I think the node version on `ci` needs to be updated?

```bash
coltonshaw@Coltons-MacBook-Pro webapp % export NODE_OPTIONS=--openssl-legacy-provider
coltonshaw@Coltons-MacBook-Pro webapp % npm run build                                

> build
> webpack --mode=production

asset main.js 23.6 KiB [emitted] [minimized] (name: main)
orphan modules 1.04 KiB [orphan] 2 modules
runtime modules 221 bytes 1 module
modules by path ./node_modules/core-js/internals/*.js 54 KiB
  ./node_modules/core-js/internals/export.js 2.54 KiB [built] [code generated]
  ./node_modules/core-js/internals/global.js 590 bytes [built] [code generated]
  ./node_modules/core-js/internals/function-call.js 206 bytes [built] [code generated]
  ./node_modules/core-js/internals/is-object.js 154 bytes [built] [code generated]
  ./node_modules/core-js/internals/is-callable.js 160 bytes [built] [code generated]
  ./node_modules/core-js/internals/get-built-in.js 358 bytes [built] [code generated]
  ./node_modules/core-js/internals/engine-v8-version.js 850 bytes [built] [code generated]
  ./node_modules/core-js/internals/a-callable.js 386 bytes [built] [code generated]
  ./node_modules/core-js/internals/well-known-symbol.js 1.03 KiB [built] [code generated]
  + 88 modules
./src/index.tsx + 2 modules 1.46 KiB [built] [code generated]
./node_modules/core-js/modules/es.promise.js 14 KiB [built] [code generated]
webpack 5.67.0 compiled successfully in 1764 ms
coltonshaw@Coltons-MacBook-Pro webapp % git commit -m "Fixed final jest issue"
[package-update a6e00ec] Fixed final jest issue
 2 files changed, 11 insertions(+), 3 deletions(-)
coltonshaw@Coltons-MacBook-Pro webapp % git push
Enumerating objects: 9, done.
Counting objects: 100% (9/9), done.
Delta compression using up to 12 threads
Compressing objects: 100% (5/5), done.
Writing objects: 100% (5/5), 572 bytes | 572.00 KiB/s, done.
Total 5 (delta 4), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (4/4), completed with 4 local objects.
To https://github.com/coltoneshaw/mattermost-plugin-starter-template.git
   102a40f..a6e00ec  package-update -> package-update
coltonshaw@Coltons-MacBook-Pro webapp % 
```

## Jest

```bash
coltonshaw@Coltons-MacBook-Pro webapp % npm run test

> test
> jest --forceExit --detectOpenHandles --verbose

 PASS  src/manifest.test.tsx
  ✓ Plugin manifest, id and version are defined (4 ms)
  ✓ Plugin id and version are defined (2 ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        3.775 s, estimated 4 s
Ran all test suites.
coltonshaw@Coltons-MacBook-Pro webapp % 
```

## Lint

```bash
coltonshaw@Coltons-MacBook-Pro webapp % npm run lint

> lint
> eslint --ignore-pattern node_modules --ignore-pattern dist --ext .js --ext .jsx --ext tsx --ext ts . --quiet --cache

coltonshaw@Coltons-MacBook-Pro webapp % 
```

**Comment by hanzei** [comment]
CI does indeed use node 14 (https://hub.docker.com/layers/circleci/golang/1.16.0-node/images/sha256-3c826afdfe01aaf82c1bd1a66e723ad7cc1ec1dea67678322eda8751a48cec8a?context=explore). A newer version of the CI image uses node 16.13.1 (https://hub.docker.com/layers/circleci/golang/1.17-node/images/sha256-13cbe95655acf5f0fd4cacbb33211ed70538b041e94f7d370926dd7042b056bd?context=explore) Is that enough?


Other question: Do you know newer node versions can run the existing code? That would be a requirement as all plugins use the same CI version.

**Comment by coltoneshaw** [comment]
Node 16.13.1 does work for this when i tested the other day. Right now i'm hitting some `sha` validation issues with `react-bootstrap`. I'll play with that and maybe just build the `package-lock.json` for 16.13.

For newer node versions running the older plugin code, it almost entirely depends on the packages installed it seems. Example, the old `mattermost-webapp` package that was imported in a lot of plugins `"github:mattermost/mattermost-webapp#23f5f93d9f12a7e2b5623e5cee6814366abd9a0f"` seems to be dated an includes `node-sass` that isn't compatible with newer node versions. Ideally, those packages are _also_ updated in those plugins.

Side note, I realize I moved the redux link to use the `mattermost-webapp` package but didn't remove the dependency, let me redo that.

**Comment by coltoneshaw** [comment]
Okay, rebuild the `package-lock.json` for node 16.14 and npm 8. Figured that was a better long-term option since node 16 is the LTS version. Everything runs locally just fine. Additionally, removed the `mattermost-redux` package from the plugin.

I did a test with node 17 and the build fails. If the `package-lock` is removed, the node packages will install properly.

```
coltonshaw@Coltons-MacBook-Pro webapp % npm run build

> build
> webpack --mode=production

asset main.js 23.6 KiB [compared for emit] [minimized] (name: main)
orphan modules 1.04 KiB [orphan] 2 modules
runtime modules 221 bytes 1 module
modules by path ./node_modules/core-js/internals/*.js 54 KiB
  ./node_modules/core-js/internals/export.js 2.54 KiB [built] [code generated]
  ./node_modules/core-js/internals/global.js 590 bytes [built] [code generated]
  ./node_modules/core-js/internals/function-call.js 206 bytes [built] [code generated]
  ./node_modules/core-js/internals/is-object.js 154 bytes [built] [code generated]
  ./node_modules/core-js/internals/is-callable.js 160 bytes [built] [code generated]
  ./node_modules/core-js/internals/get-built-in.js 358 bytes [built] [code generated]
  ./node_modules/core-js/internals/engine-v8-version.js 850 bytes [built] [code generated]
  ./node_modules/core-js/internals/a-callable.js 386 bytes [built] [code generated]
  ./node_modules/core-js/internals/well-known-symbol.js 1.03 KiB [built] [code generated]
  + 88 modules
./src/index.tsx + 2 modules 1.46 KiB [built] [code generated]
./node_modules/core-js/modules/es.promise.js 14 KiB [built] [code generated]
webpack 5.67.0 compiled successfully in 1913 ms
coltonshaw@Coltons-MacBook-Pro webapp % npm run lint

> lint
> eslint --ignore-pattern node_modules --ignore-pattern dist --ext .js --ext .jsx --ext tsx --ext ts . --quiet --cache

coltonshaw@Coltons-MacBook-Pro webapp % npm run test

> test
> jest --forceExit --detectOpenHandles --verbose

watchman warning:  Recrawled this watch 17 times, most recently because:
MustScanSubDirs UserDroppedTo resolve, please review the information on
https://facebook.github.io/watchman/docs/troubleshooting.html#recrawl
To clear this warning, run:
`watchman watch-del '/Users/coltonshaw/go/src/github.com/mattermost/mattermost-plugin-starter-template' ; watchman watch-project '/Users/coltonshaw/go/src/github.com/mattermost/mattermost-plugin-starter-template'`

 PASS  src/manifest.test.tsx
  ✓ Plugin manifest, id and version are defined (5 ms)
  ✓ Plugin id and version are defined (6 ms)

Test Suites: 1 passed, 1 total
Tests:       2 passed, 2 total
Snapshots:   0 total
Time:        3.296 s, estimated 4 s
Ran all test suites.
coltonshaw@Coltons-MacBook-Pro webapp %
```

File: README.md
Comment (mickmister) [inline README.md:7]: Nit: spacing

```suggestion
This template requires node v16 and npm v8.

```

File: package-lock.json
Comment (mickmister) [inline package-lock.json:6]: This file can be removed

File: webapp/src/index.tsx
Comment (mickmister) [inline webapp/src/index.tsx:3]: Are we able to successfully import and use functional exports from `mattermost-redux`, such as selectors?

File: webapp/package.json
Comment (mickmister) [inline webapp/package.json:3]: Wondering if this change is required

**Review by mickmister — APPROVED** [review]
Changes LGTM :+1:

I've added @hmhealey to review the specifics of importing `mattermost-webapp`

File: package-lock.json
Comment (coltoneshaw) [reply package-lock.json:6]: We want to remove the whole package-lock.json?

File: webapp/src/index.tsx
Comment (coltoneshaw) [reply webapp/src/index.tsx]: I just pushed a commit that allows for this to be imported that way. Had to figure out all the alias mess with eslint, webpack, and typescript.

File: webapp/package.json
Comment (coltoneshaw) [reply webapp/package.json:3]: Might not be required, I was working to clean up some of the webpack config imports and such and that may be leftover from that. We can remove it if you'd like!

File: webapp/package.json
Comment (hmhealey) [inline webapp/package.json:66]: In particular, is this going to cause npm to try to pick up new versions of the web app whenever someone does an npm install like it likes to do for non-exact versions of other dependencies?

**Review by hmhealey — COMMENTED** [review]
What's the motivation for removing the mattermost-redux dependency and switching everything to the web app? I know we generally want to keep some stuff up to date, but by updating it, I'm worrying we're potentially opening ourselves up to more risky and duplicated code in the plugins that is more likely to break going forward. At least by leaving it as-is, we're keeping the same status quo.

File: webapp/package.json
Comment (coltoneshaw) [reply webapp/package.json:66]: We can definitely do a specific version hash to pull based on the plugin now. I did this because I kept hitting issues downloading the packages and this seemed to be one of the only solutions that worked.

**Comment by coltoneshaw** [comment]
> What's the motivation for removing the mattermost-redux dependency and switching everything to the web app? I know we generally want to keep some stuff up to date, but by updating it, I'm worrying we're potentially opening ourselves up to more risky and duplicated code in the plugins that is more likely to break going forward. At least by leaving it as-is, we're keeping the same status quo.

The idea here is from @hanzei / @mickmister in order to remove excess packages. Now since it's included in the current web app packages, it seemed like we could just pull from there and keep it up to date. Especially now with the Mattermost-redux repo being archived.

I think this would only really be used for people making new plugins, and the old plugins that exist would have to handle the upgrade how they see fit. Is that what you meant?

File: webapp/package.json
Comment (hmhealey) [reply webapp/package.json:66]: Was it the one about the integrity check being wrong? That'd happen for every dependency pulled directly from GitHub unfortunately. If that's the case, we've done some similar fixes [in the web app](https://github.com/mattermost/mattermost-webapp/pull/9617) and [other plugins](https://github.com/mattermost/mattermost-plugin-calls/pull/3).

I think doing that and depending on a specific version would be safer since there's no risk of the web app being accidentally updated by NPM and breaking something if any of the functions used were to change or be removed

**Comment by coltoneshaw** [comment]
That's fair and I think that shouldn't pose a problem. There aren't any other imports from the web app that are needed at this time (that I can see). 

Is the `mattermost-redux` package updated regularly? That's the only concern I could think of. Since we moved to v6 Mattermost and redux says v5.31, but those could be entirely unrelated versioning.

**Comment by coltoneshaw** [comment]
Wait, we do import the tests from the `webapp`

File: webapp/package.json
Comment (coltoneshaw) [reply webapp/package.json:66]: That's fair. We could essentially tie this dependency to the ESR, if we wanted to leave the `mattermost-webapp` package in here.

**Comment by hanzei** [comment]
@hmhealey One reason to import `mattermost-redux` from `webapp` is so that plugin can get an up-to-date version. The latest redux version is a year old. Some plugin had to copy code from the webapp just so they can start using newer functionality.

FYI: There are plugins who already import the webapp directly: https://github.com/mattermost/mattermost-plugin-incident-management/blob/43ae217342d86a7cf117404a75bf8a67d18787c7/webapp/src/actions.ts#L11

File: package-lock.json
Comment (mickmister) [reply package-lock.json:6]: I think you may have accidentally run `npm i` in the root of the repository, which created this file. `webapp/package-lock.json` should be kept, but this file can be removed.

Comment (coltoneshaw) [reply package-lock.json:6]: AHAH! I did keep doing that. I'll go ahead and delete it.

**Comment by hmhealey** [comment]
With it being out of date, I know that there's some new features certain plugins want to use, but I don't think that needs to be part of the starter template since most plugins won't need them. The old version of mattermost-redux includes most common types and things that users would want to do since a lot of the standard stuff around users, posts, and channels haven't changed, so I don't think there's much hard staying with something old and stable. That said, this isn't my area of ownership, so I don't want to block this for that reason any more.

I am 3/5 against removing the lock file though since that will result in dependencies of dependencies changing unexpectedly which could break stuff randomly. I think specifying a fixed commit hash in the package.json and using the method I linked [here](https://github.com/mattermost/mattermost-plugin-starter-template/pull/163#discussion_r799618496) to skip the buggy integrity check would be a safer way to do that.

**Comment by hanzei** [comment]
> With it being out of date, I know that there's some new features certain plugins want to use, but I don't think that needs to be part of the starter template since most plugins won't need them. The old version of mattermost-redux includes most common types and things that users would want to do since a lot of the standard stuff around users, posts, and channels haven't changed, so I don't think there's much hard staying with something old and stable. That said, this isn't my area of ownership, so I don't want to block this for that reason any more.

I would like to allow for plugin devs picking up the starter-template to not care about the future of there build tools and use the "latest & greatest". This way we also ensure that also advanced plugins can use advanced features.

**Comment by hmhealey** [comment]
Whoops. I submitted these in the wrong order 😛

That makes sense. I still have my doubts about whether most plugins will need the new stuff, but that's not my call to make anyway. It would be nice if we could have that only affect our code and have it pinned at setup time so that versions don't change down the line, but I guess that would require something more scripted like `create-react-app`.

Also, sorry, I think I got confused a bit last time. I saw the commit for "Remove root package-lock.json", but I didn't realize that was cleaning up an accidental file. I thought it was removing the actual `package-lock.json` in the webapp folder

**Comment by coltoneshaw** [comment]
@hmhealey - Okay, so final resolution for this is to leave it without the redux dependency, and peg the release to an ESR commit for our webapp repo. Is that correct?

@hanzei - I think the CI still needs to have the node version updated to move forward.

**Comment by hmhealey** [comment]
Yeah, we're continuing with the mattermost-webapp dependency, but I don't remember where we ended up on setting the version. No matter what version we set it to, it'll remain unchanged as long as the person using this template doesn't update it, so I don't have any strong feelings particularly about any specific commit to set it to

**Comment by coltoneshaw** [comment]
Okay. My thought is that if we leave it open-ended then we run the chance of a plugin dev continuing to update to the latest master branch, without much rhyme or reason. That feels like it could pose an issue. 

Wondering if we peg it to the ESR releases, and just update the template here twice a year, and add it to the plugin docs that devs should do the same or follow the plugin channel.

Or is there any chance of making this an NPM package?

@hanzei / @mickmister - You folks are the best advisors on this.

**Comment by JulienTant** [comment]
Hey guys, is there anything I can do to help get this merged?

**Comment by coltoneshaw** [comment]
Right now the only thing left is to update the the node version in the tests to have it pass. @hanzei Can you help with that?

I can add a commit to peg this to v6.3 Mattermost web app for right now and maybe a doc on updating it. @mickmister Let me know how you feel about that.

File: webapp/package.json
Comment (mickmister) [reply webapp/package.json:66]: @coltoneshaw I'm thinking we should import the `mattermost-webapp` repo from a specific git commit and follow @hmhealey's strategy of avoiding the integrity check

**Comment by mickmister** [comment]
@hanzei I'd like to get these changes passing with CI, and include the same changes to the e2e tests in the Apps plugin https://github.com/mattermost/mattermost-plugin-apps/pull/323. What changes need to be done to the circleci config in the starter template repo to update the npm version? Is there documentation I can reference to figure this out? I'm not knowledgeable on how our circleci org setup works.

**Comment by hmhealey** [comment]
FYI, when we next update the web app package, we're going to start running into issues with importing parts of the webapp because of the monorepo. We'll likely going to need to make similar changes to the ones I explain [here](https://community.mattermost.com/core/pl/wp6pubzympg4fxa3crxgk8zk1a), although those will be changing a few times over the coming weeks, so perhaps it would be a good idea to wait until those stabilize.

File: webapp/package.json
Comment (mickmister) [inline webapp/package.json:76]: I'm wondering if we can remove this dependency. It's only used for setting up tests, and even then nothing is really Mattermost-specific in that setup.

**Review by mickmister — COMMENTED** [review]
CI is now updated to support using a plugin project's `.nvmrc` file to build, so I'm thinking this PR is unblocked by CI now. How can I help push this PR through?

**Comment by mickmister** [comment]
CI seems to not like when the `eslintrc.json` file contains `./webapp/tsconfig.json` as the path to `tsconfig.json`, while `./tsconfig.json`. Unfortunately this makes the `vscode-eslint` plugin not work properly.

The latest commit adds a `.vscode/settings` file, to tell the eslint plugin to use the `webapp` directory. This is the only way I've gotten it work https://github.com/microsoft/vscode-eslint/issues/1289#issuecomment-873376465

**Comment by mickmister** [comment]
I temporarily have the webapp setup not being imported in tests, which likely means writing unit tests for React components right now. I couldn't resolve the infamous `Cannot use import statement outside a module` error this time around. Upgrading `jest` made it so I could import things, but I ran into other compatibility issues.

I suggest we get this merged, and fix the test setup as a follow-up task. What do you think @coltoneshaw @DHaussermann?

**Comment by coltoneshaw** [comment]
@mickmister - I'm happy to get this merged whenever you're ready! Let me know what you might need from me

**Comment by mickmister** [comment]
@hmhealey Can you take a look at some of the changes in the last few commits I've added here? I'm wondering if there's a more proper way to have this set up, the given the use cases of the plugin projects.

**Comment by mickmister** [comment]
CI is finally passing :tada: along with VSCode being aware of eslint and typescript

**Comment by mickmister** [comment]
cc @hanzei Let's get this merged so we have a stable starter template

File: .nvmrc
Comment (hanzei) [inline .nvmrc:1]: Why not 
```suggestion
16.13.1
```

File: webapp/package.json
Comment (hanzei) [reply webapp/package.json:77]: 1/5 it might be worth keeping it around as it makes it easier for dev touse it later if the need it.

File: webapp/package.json
Comment (hanzei) [inline webapp/package.json:77]: Not locked version?

File: webapp/src/index.tsx
Comment (hanzei) [inline webapp/src/index.tsx:3]: :100:

**Review by hanzei — APPROVED** [review]
Neat :rocket:

**Comment by hanzei** [comment]
@coltoneshaw Please let me know what you think about my non-blocking comments above and then let's merge the PR :rocket:

File: webapp/package.json
Comment (mickmister) [reply webapp/package.json:77]: I agree, we should have a specific commit/version here

**Comment by mickmister** [comment]
@hanzei I think Colton is handing off the PR to our team, so I think any further update should just be done by us, and not wait for his approval or intervention

**Comment by coltoneshaw** [comment]
> @hanzei I think Colton is handing off the PR to our team, so I think any further update should just be done by us, and not wait for his approval or intervention

Feel free to do whatever is needed to have this merged! I have a few other things on my plate and haven't been paying attention to this at the moment. I see the asks are tiny, so if needed I can commit them today so we can move this along.

File: webapp/package.json
Comment (mickmister) [reply webapp/package.json:77]: @hanzei Any version in particular you think we should target? I'm not sure what the implications would be, though I would favor more updated over old version

Comment (hanzei) [reply webapp/package.json:77]: It's only used in https://github.com/mattermost/mattermost-plugin-starter-template/blob/876ac24bd3e1c3270d7176b12549a2451266d948/webapp/tests/setup.tsx#L4, and not in production code. 1/5 we just use the latest.

**Comment by hanzei** [comment]
@mickmister 1/5 let's merge this PR as it is and clean up the webapp imports later. We can properly also remove the `mattermost-redux` rules from `tsconfig` and `webpack`.

File: webapp/package.json
Comment (coltoneshaw) [reply webapp/package.json:77]: I'd always target the ESR and update it regularly to be the latest ESR. Gives us the best ability to keep this in supported range.

**Comment by coltoneshaw** [comment]
@hanzei - `mattermost-redux` alias I added to make it easier to import the code for redux. Without it the imports are kind of ugly / confusing. It's the same thing that Calls does - https://github.com/mattermost/mattermost-plugin-calls/blob/main/webapp/tsconfig.json

File: webapp/package.json
Comment (mickmister) [reply webapp/package.json:77]: @hanzei @coltoneshaw FYI I ran into these errors when trying to pin the dep to a specific commit, trying several `7.x` releases:

```
webapp [package-update●] npm i
npm WARN deprecated @floating-ui/react-dom-interactions@0.9.1: Package renamed to @floating-ui/react
npm ERR! code 1
npm ERR! path /Users/koch/mattermost/mattermost-plugin-starter-template/webapp/node_modules/@mattermost/webapp
npm ERR! command failed
npm ERR! command sh -c npm run build --workspaces
npm ERR! > @mattermost/client@7.1.0 build
npm ERR! > tsc --build --verbose
npm ERR! 
npm ERR! 12:01:40 PM - Projects in this build: 
npm ERR!     * ../types/tsconfig.json
npm ERR!     * tsconfig.json
npm ERR! 
npm ERR! 12:01:40 PM - Project '../types/tsconfig.json' is out of date because output file '../types/lib/admin.js' does not exist
npm ERR! 
npm ERR! 12:01:40 PM - Building project '/Users/koch/mattermost/mattermost-plugin-starter-template/webapp/node_modules/@mattermost/webapp/packages/types/tsconfig.json'...
npm ERR! 
npm ERR! 12:01:43 PM - Project 'tsconfig.json' is out of date because output file 'lib/client4.js' does not exist
npm ERR! 
npm ERR! 12:01:43 PM - Building project '/Users/koch/mattermost/mattermost-plugin-starter-template/webapp/node_modules/@mattermost/webapp/packages/client/tsconfig.json'...
npm ERR! 
npm ERR! 
npm ERR! > @mattermost/types@7.1.0 build
npm ERR! > tsc --build --verbose
npm ERR! 
npm ERR! 12:01:45 PM - Projects in this build: 
npm ERR!     * tsconfig.json
npm ERR! 
npm ERR! 12:01:46 PM - Project 'tsconfig.json' is up to date because newest input 'src/utilities.ts' is older than oldest output 'lib/audits.js'
npm ERR! 
npm ERR! 
npm ERR! > build
npm ERR! > rollup -c
npm ERR! sh: rollup: command not found
npm ERR! npm ERR! Lifecycle script `build` failed with error: 
npm ERR! npm ERR! Error: command failed 
npm ERR! npm ERR!   in workspace: @mattermost/components 
npm ERR! npm ERR!   at location: /Users/koch/mattermost/mattermost-plugin-starter-template/webapp/node_modules/@mattermost/webapp/packages/components

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/koch/.npm/_logs/2023-02-21T17_01_47_242Z-debug.log
```

Comment (mickmister) [reply webapp/package.json:77]: This error in particular references `rollup: command not found`. I'm not sure if we want to copy the build tooling from the webapp that builds each individual package there. cc @hmhealey

**Pull request merged**
